### ISSUE TITLE
Suggestion: false positive of `no-invalid-this` rule on event handlers

### ISSUE THREAD

**Issue by Antonius-S** [eslint/eslint#11349]
**What rule do you want to change?**
no-invalid-this
**Does this change cause the rule to produce more or fewer warnings?**
fewer
**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior
**Please provide some example code that this change will affect:**

```js
'use strict';
process.on('exit', function(){ console.log(this.argv); });
```
the same with `addListener()`
**What does the rule currently do for this code?**
Rule complains on invalid `this`
**What will the rule do after it's changed?**
Rule must consider `EventEmitter` interface (`on()` and `addListener()` methods) and allow `this` in event handlers
**Are you willing to submit a pull request to implement this change?**
I've no required knowledge

**Comment by g-plane** [comment]
Repro link: https://eslint.org/demo/#eyJ0ZXh0IjoiJ3VzZSBzdHJpY3QnO1xyXG5wcm9jZXNzLm9uKCdleGl0JywgZnVuY3Rpb24oKXsgY29uc29sZS5sb2codGhpcy5hcmd2KTsgfSk7XHJcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1pbnZhbGlkLXRoaXMiOjJ9LCJlbnYiOnt9fX0=

**Comment by g-plane** [comment]
While we may not intend to change to core rules, we can update the documentation for the section "When Not to Use" of this rule. Should we add a "documentation" label for this issue?

**Comment by aladdin-add** [comment]
I'm open to discuss, let's just see other members thoughts! 😄

**Comment by mysticatea** [comment]
This is by design.

At that time (https://github.com/eslint/eslint/pull/2857), we had considered `allowInCallback` option, but it had been rejected.

You can use a workaround:

```js
'use strict';

process.on('exit', function(){ console.log(process.argv); });
// OR 
process.on('exit', /** @this {Process} */ function(){ console.log(this.argv); });
```

----

> I'll redirect this suggestion to eslint-plugin-node

https://github.com/mysticatea/eslint-plugin-node/issues/4 is a long-living issue. But that issue is not high priority to me because `this` in callbacks is not popular thing after arrow functions have prevailed.

**Comment by Antonius-S** [comment]
> [mysticatea/eslint-plugin-node#4](https://github.com/mysticatea/eslint-plugin-node/issues/4) is a long-living issue. But that issue is not high priority to me because `this` in callbacks is not popular thing after arrow functions have prevailed.

This is rather strange to me. Using `this` in callbacks for sockets, streams, files are very usual in Node. Moreover the need of pointer to a caller prevents many callbacks from converting to arrow functions. Especially when callbacks are tiny so declaring them as separate functions is excessive. F.i.,
```js
fileStream.on('close', function() {this.opened = false})
```
or
```js
socket.on('error', function() { reportError(this) })
```

**Comment by platinumazure** [comment]
@g-plane What sort of note would you want added in documentation? Not opposed in the least, just trying to get some more context.

Agreed that this is something we don't want to change in ESLint core.

**Comment by g-plane** [comment]
Just saying something like this: "if you are using EventEmitter in Node.js".

**Comment by mysticatea** [comment]
> This is rather strange to me.

Of course, your contribution is welcome.

> Just saying something like this: "if you are using EventEmitter in Node.js".

I'd like to be more specific a bit. How about "if you want to use `this` in the callback of environment-specific classes such as `EventTarget` or `EventEmitter`."?
Actually, `EventTarget`, the standard class which handles events in browsers, passes `this` to event listeners.

**Comment by Antonius-S** [comment]
> Of course, your contribution is welcome.
I'd like to, but interface is too complex and I didn't found docs for ASTTree node object. AFAIU the stack stores a function node and all is needed is a name check but I've no clue how to implement this.
